### DISCUSSION TITLE
STL support.

### DISCUSSION THREAD

**Discussion by jfjlaros** [jfjlaros/arduino-simple-rpc#24] [Ideas]
Continuation of #23.

**Comment by gpantazis** [comment]
Ok so due to the nature of the job, I can't share my actual code using `std::string`. 

Instead I have attached a minimal project, as well as the edits to each of the necessary files.

### Minimal Project
`test)stl_str.ino`:
```
#include <string>

#include "simpleRPC.h"

void setup() {
    Serial.begin(9600);
    while(!Serial) {}
}

std::string greet(std::string person)
{
    return "Hello, "  + person + "\n";
}

void loop() {
    interface(Serial,
              greet, "greet:");
}
```

### Modified Files
Modified files are below. Note that the changes in all of them mirror the functions for `String`, the Arduino built-in `std::string` equivalent. For `std::vector` and `std::array` you would likely follow the definitions for `Vector`

`types.tcc`:
```
#include <string>
....

/*! \ingroup types
 * \copydoc rpcTypeOf(Stream&, bool) */
inline void rpcTypeOf(Stream& io, std::string&) {
  rpcPrint(io, "s");
}
```

`write.tcc`
```
#include <string>
...
/*! \ingroup write
 * \copydoc rpcWrite(Stream&, T*) */
inline void rpcWrite(Stream& io, std::string* data) {
  rpcPrint(io, *data, '\0');
}
```

`print.tcc`:
```
#include <string>
...
/*! \ingroup print
 * \copydoc rpcPrint(Stream&, T) */
inline void rpcPrint(Stream& io, std::string& data) {
  rpcPrint(io, data.c_str());
}
```

`read.tcc`:
```
#include <string>
...
/*! \ingroup read
 * \copydoc rpcRead(Stream&, T*) */
inline void rpcRead(Stream& io, std::string* data) {
  char character;

  rpcRead(io, &character);

  while (character) {
    *data += character;
    rpcRead(io, &character);
  }
}
```

### Build Instructions
This is where for non-Teensy boards I am not too clear on. I checked through the base Arduino avr folder and there are no C++ libraries. Downloading support for the arm boards (Arduino Due) seems to yield C++ libs as part of their ARM compiler , but not `-lstdc++_nano`. This may be because they use an older version of gcc (4.8.3-2014q1). That said `-lstdc++` is available. This coupled with `-specs=nano.specs` should have the same effect (on code size) as `-lstdc++_nano`.

So I got the code above to compile for Arduino Due by editing this file, `C:\Users\<USER>\Local\ArduinoXX\Packages\arduino\hardware\sam\1.6.12\platform.txt`, and appending `-lstdc++`  and `-specs=nano.specs` to the end of this line (see code block below). I don't know how to do this using `arduino-cli`, `boards.txt` or `platform.local.txt`
```
...
## Combine gc-sections, archives, and objects
recipe.c.combine.pattern="{compiler.path}{compiler.c.elf.cmd}" -mcpu={build.mcu} -mthumb {compiler.c.elf.flags} "-T{build.variant.path}/{build.ldscript}" "-Wl,-Map,{build.path}/{build.project_name}.map" {compiler.c.elf.extra_flags} -o "{build.path}/{build.project_name}.elf" "-L{build.path}" -Wl,--cref -Wl,--check-sections -Wl,--gc-sections -Wl,--entry=Reset_Handler -Wl,--unresolved-symbols=report-all -Wl,--warn-common -Wl,--warn-section-align -Wl,--start-group {compiler.combine.flags} {object_files} "{build.variant.path}/{build.variant_system_lib}" "{build.path}/{archive_file}" -Wl,--end-group -lm -lgcc -lstdc++ -specs=nano.specs
...
```
Note that adding `-specs=nano.specs` to the end of the line above drastically reduces the size of the resulting sketch from 74592 bytes to 12488 bytes (!).

**Reply by jfjlaros** [reply]
I managed to compile your sketch using `arduino-cli`. Below are my notes.

# Installation

Install the core for ARM boards.

    arduino-cli core install arduino:sam

Edit `~/arduino15/packages/arduino/hardware/sam/1.6.12/platform.txt`, add `-lstdc++ -specs=nano.specs` to the end of the `recipe.c.combine.pattern` line.

# Example sketch

Configure the sketch for the right core.

    arduino-cli board attach arduino:sam:arduino_due_x

Content of `stl.ino`:

```c++
#include "stl.h"


std::string greet(std::string person) {
  return "Hello, " + person + "\n";
}


void setup() {
  Serial.begin(9600);
}

void loop() {
  interface(Serial, greet, "");
}
```

Content of `stl.h`:

```c++
#include <string>

#include <simpleRPC.h>


inline void rpcTypeOf(Stream& io, std::string&) {
  rpcPrint(io, "s");
}

inline void rpcWrite(Stream& io, std::string* data) {
  rpcPrint(io, *data, '\0');
}

inline void rpcPrint(Stream& io, std::string& data) {
  rpcPrint(io, data.c_str());
}

inline void rpcRead(Stream& io, std::string* data) {
  char character;

  rpcRead(io, &character);

  while (character) {
    *data += character;
    rpcRead(io, &character);
  }
}
```

# Compilation

Compile the sketch.

    arduino-cli compile

**Comment by jfjlaros** [comment]
To use the STL data types, we could offer an excerpt of the notes above as well as an alternative include line, e.g.,

```c++
#include <simpleRPC_STL.h>
```

That way we would have support for both the custom data types as well as the ones in the STL.

**Comment by jfjlaros** [comment]
The Arduino library now has experimental [STL support](https://github.com/jfjlaros/simpleRPC/tree/stl). I have no idea whether it works since I have no way of testing it.

You can find the example sketch [here](https://github.com/jfjlaros/simpleRPC/blob/stl/examples/stl/stl.ino).

**(edit)**

Apparently the ESP32 core has STL support out of the box, so I can do some testing after all. In this [sketch](https://github.com/jfjlaros/simpleRPC/blob/stl/examples/esp32_stl/esp32_stl.ino) `std::string` and `std::vector` seem to be working.
